### PR TITLE
Delay opening redirected files until execing cmd

### PR DESCRIPTION
**Pull request by jonludlam**

### CODE REVIEW

File: src/action_exec.ml [hunk 1]
```diff
@@ -44,7 +40,7 @@ let exec_echo stdout_to str =
   Fiber.return
     (match stdout_to with
      | None -> print_string str; flush stdout
-     | Some (_, oc) -> output_string oc str)
+     | Some fn -> Io.with_file_out fn ~f:(fun oc -> output_string oc str))
```
Comment (rgrinberg) [inline src/action_exec.ml]: This can be simplified to use `Io.write_file` now.

File: src/action_exec.ml [hunk 2]
```diff
@@ -195,16 +179,15 @@ let rec exec t ~ectx ~dir ~env ~stdout_to ~stderr_to =
     Fiber.return ()
 
 and redirect outputs fn t ~ectx ~dir ~env ~stdout_to ~stderr_to =
-  let oc = Io.open_out fn in
-  let out = Some (fn, oc) in
+  let abs = Path.to_absolute fn in
```
Comment (rgrinberg) [inline src/action_exec.ml]: Why this is necessary deserved a comment.

**Review by rgrinberg — CHANGES_REQUESTED** [review]
Looks good. A `CHANGES` entry is appropriate as well.

File: src/action_exec.ml (hunk 2 above)
Comment (jonludlam) [reply src/action_exec.ml]: Mostly just paranoia in the presence of 'Chdir' as an action. Dune itself still builds if we don't resolve the path at this point, but I didn't want to assume that everything else will. If you think it's unnecessary I'll remove it, or we can keep it in and stick a comment there.

Comment (rgrinberg) [reply src/action_exec.ml]: A comment would be useful here I think. Thanks.

**Review by rgrinberg — APPROVED** [review]

**Comment by diml** [comment]
This doesn't quite work, for instance this action would now not produce the expected result:

```
(with-stdout-to file
 (progn
  (echo foo)
  (echo bar)))
```

**Pull request merged**